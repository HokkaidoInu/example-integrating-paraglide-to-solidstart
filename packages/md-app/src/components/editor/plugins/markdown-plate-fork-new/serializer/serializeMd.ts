import type { Descendant, SlateEditor } from '@udecode/plate';

import remarkStringify from 'remark-stringify';
import { type Plugin, unified } from 'unified';

import type { AllowNodeConfig, NodesConfig } from '../MarkdownPlugin';
import type { MdRoot } from '../mdast';
import type { TNodes } from "../node-rules";

import { convertNodesSerialize } from './convertNodesSerialize';
import { getMergedOptionsSerialize } from './utils/getMergedOptions';

export type SerializeMdOptions = {
  allowedNodes?: NodesConfig;
  allowNode?: AllowNodeConfig;
  disallowedNodes?: NodesConfig;
  editor?: SlateEditor;
  nodes?: TNodes | null;
  remarkPlugins?: Plugin[];
  value?: Descendant[];
};

/** Serialize the editor value to Markdown. */
export const serializeMd = (
  editor: SlateEditor,
  options?: Omit<SerializeMdOptions, 'editor'>
) => {
  const mergedOptions = getMergedOptionsSerialize(editor, options);

  const { remarkPlugins, value } = mergedOptions;

  const toRemarkProcessor = unified()
    .use(remarkPlugins ?? [])
    .use(remarkStringify, {
      // Configure remark-stringify to handle MDX JSX elements
      handlers: {
        mdxJsxTextElement: (node, _, state, info) => {
          
          const attrs = node.attributes
            .map((attr: any) => `${attr.name}="${attr.value}"`)
            .join(' ');
          const attrStr = attrs ? ` ${attrs}` : '';

          // Handle underline elements specifically
          const content = node.children[0]?.value || '';
          return `<${node.name}${attrStr}>${content}</${node.name}>`;
        },
      },
    });
  const mdast = slateToMdast({
		children: value!,
		options: mergedOptions,
	});

  return toRemarkProcessor.stringify(mdast);
};

const slateToMdast = ({
  children,
  options,
}: {
  children: Descendant[];
  options: SerializeMdOptions;
}): MdRoot => {
  const ast = {
    children: convertNodesSerialize(children, options) as MdRoot['children'],
    type: 'root',
  } as MdRoot;
  return ast;
};
