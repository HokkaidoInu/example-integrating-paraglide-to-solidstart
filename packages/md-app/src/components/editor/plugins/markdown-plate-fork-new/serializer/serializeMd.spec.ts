import { createTestEditor } from '../__tests__/createTestEditor';
import { testValue } from '../__tests__/testValue';
import { serializeMd } from './serializeMd';
const editor = createTestEditor();

describe('serializeMd', () => {
  it('should serialize a simple paragraph', () => {
    const result = serializeMd(editor as any, { value: testValue });
    expect(result).toMatchSnapshot();
  });

  describe('fixures', () => {
    // https://github.com/inokawa/remark-slate-transformer/issues/44
    it('should serialize marks with trailing whitespace', () => {
      const slateNodes = [
        {
          children: [
            {
              text: 'Make text ',
            },
            {
              bold: true,
              text: 'bold',
            },
            {
              text: ', ',
            },
            {
              italic: true,
              text: 'italic',
            },
            {
              text: ', ',
            },
            {
              text: 'underlined',
              underline: true,
            },
            {
              text: ', or apply a ',
            },
            {
              bold: true,
              italic: true,
              text: 'combination',
              underline: true,
            },
            {
              text: ' of these styles for a visually striking effect.',
            },
          ],
          type: 'p',
        },
      ];
      const result = serializeMd(editor as any, { value: slateNodes });
      expect(result).toMatchSnapshot();
    });

    // https://github.com/inokawa/remark-slate-transformer/issues/90
    it('Combination of bold and italic at the start of a block losing formatting', () => {
      const slateNodes = [
        {
          children: [
            {
              bold: true,
              italic: true,
              text: 'Italic',
            },
            {
              bold: true,
              text: ' in a bold paragraph',
            },
          ],
          type: 'p',
        },
        {
          children: [
            {
              bold: true,
              text: 'This is an ',
            },
            {
              bold: true,
              italic: true,
              text: 'Italic',
            },
            {
              bold: true,
              text: ' in a bold paragraph',
            },
          ],
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, { value: slateNodes });
      expect(result).toMatchSnapshot();
    });

    // https://github.com/inokawa/remark-slate-transformer/issues/145
    it('should serialize a value with inline code', () => {
      const slateNodes = [
        {
          children: [
            { bold: true, code: true, italic: true, text: 'Inline code' },
          ],
          type: 'p',
        },
        {
          children: [{ code: true, italic: true, text: 'Inline code' }],
          type: 'p',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it('should serialize a value within a block qoute as a single line', () => {
      const slateNodes = [
        {
          children: [
            { text: 'Block quote' },
            { text: ' with a new line ' },
            { bold: true, code: true, italic: true, text: ' Inline code' },
          ],
          type: 'blockquote',
        },
      ];

      expect(
        serializeMd(editor as any, { value: slateNodes })
      ).toMatchSnapshot();
    });

    it(
      String.raw`should serialize a \n within a block qoute as new paragraph`,
      () => {
        const slateNodes = [
          {
            children: [
              { text: 'Block quote' },
              { text: '\n' },
              { text: 'with a new line' },
            ],
            type: 'blockquote',
          },
        ];

        expect(
          serializeMd(editor as any, { value: slateNodes })
        ).toMatchSnapshot();
      }
    );
  });
});
