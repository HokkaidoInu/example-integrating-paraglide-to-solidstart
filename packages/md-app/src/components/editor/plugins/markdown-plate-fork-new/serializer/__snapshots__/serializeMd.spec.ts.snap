// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeMd fixures Combination of bold and italic at the start of a block losing formatting 1`] = `
"***Italic*** **in a bold paragraph**

**This is an *Italic*** **in a bold paragraph**
"
`;

exports[`serializeMd fixures should serialize a value with inline code 1`] = `
"***\`Inline code\`***

*\`Inline code\`*
"
`;

exports[`serializeMd fixures should serialize marks with trailing whitespace 1`] = `
"Make text **bold**, *italic*, <u>underlined</u>, or apply a ***<u>combination</u>*** of these styles for a visually striking effect.
"
`;

exports[`serializeMd should serialize a simple paragraph 1`] = `
"# Playground

## Comments and suggestions

Add comments to your content to provide additional conte1xt, , or collaborate with others

### Heading

* 1

  * 2

    * 3

      * 4

| 1-1 | 1-2 |
| --- | --- |
| 2-1 | 2-2 |

> blockQuote

***

normal **bold***italic*~~strikethrough~~\`code\`***combine***[link](https://example.com)$E=mc^2$

![test](https://images.unsplash.com/photo-1712688930249-98e1963af7bd?q=80\\&w=2070\\&auto=format\\&fit=crop\\&ixlib=rb-4.0.3\\&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D "test")

\`\`\`javascript
// Use code blocks to showcase code snippets
function greet() {
  console.info('Hello World!');
}
\`\`\`

$$
f(x)
$$

* [x] Todo item

* Numbered list item

Centered text

Text with custom line height

Mention: User

Date: 2024-01-01

Text with highlight and <u>underline</u> and keyboard formatting

Text with color and background color
"
`;
